openedx_learning/core/components/models.py

```diff
@@ -0,0 +1,295 @@
+    component_version = models.ForeignKey(
+        ComponentVersion, on_delete=models.RESTRICT, null=True
+    )
+
+
+class PublishedComponent(models.Model):
+    """
+    For any given Component, what is the currently published ComponentVersion.
+
+    It may be possible for a Component to exist only as a Draft (and thus not
+    show up in this table).
+    """
+
+    component = models.OneToOneField(
+        Component, on_delete=models.RESTRICT, primary_key=True
+    )
+    component_version = models.ForeignKey(
+        ComponentVersion,
+        on_delete=models.RESTRICT,
+        null=True,
+    )
+    component_publish_log_entry = models.ForeignKey(
+        ComponentPublishLogEntry,
+        on_delete=models.RESTRICT,
+    )
+
+    class Meta:
+        constraints = [
+            # Only one version of a Component is considered "published" at any
+            # given time (though other versions may be available).
+            models.UniqueConstraint(
+                fields=["component", "component_version"],
+                name="pc_uniq_component_component_version",
+            )
+        ]
+
+
+class Content(models.Model):
+    """
+    This is the most basic piece of raw content data, with no version metadata.
+
+    Content stores data in an immutable Binary BLOB `data` field. This data is
+    not auto-normalized in any way, meaning that pieces of content that are
+    semantically equivalent (e.g. differently spaced/sorted JSON) will result in
+    new entries. This model is intentionally ignorant of what these things mean,
+    because it expects supplemental data models to build on top of it.
+
+    Two Content instances _can_ have the same hash_digest if they are of
+    different MIME types. For instance, an empty text file and an empty SRT file
+    will both hash the same way, but be considered different entities.
+
+    The other fields on Content are for data that is intrinsic to the file data
+    itself (e.g. the size). Any smart parsing of the contents into more
+    structured metadata should happen in other models that hang off of Content.
+
+    Content models are not versioned in any way. The concept of versioning only
+    exists at a higher level.
+
+    Since this model uses a BinaryField to hold its data, we have to be careful
+    about scalability issues. For instance, video files should not be stored
+    here directly. There is a 10 MB limit set for the moment, to accomodate
+    things like PDF files and images, but the itention is for the vast majority
+    of rows to be much smaller than that.
+    """
+
+    # Cap item size at 10 MB for now.
+    MAX_SIZE = 10_000_000
+
+    learning_package = models.ForeignKey(LearningPackage, on_delete=models.CASCADE)
+    hash_digest = hash_field()
+
+    # Per RFC 4288, MIME type and sub-type may each be 127 chars.
+    type = models.CharField(max_length=127, blank=False, null=False)
+    sub_type = models.CharField(max_length=127, blank=False, null=False)
```

Comment:
How would you feel about `media_type` and `media_subtype`?

Reply:
Great 👍🏻

Reply:
What's the rationale for splitting media type into two separate fields? Seems a bit strange to me.

Reply:
We had some brief discussion here: https://github.com/openedx/openedx-learning/pull/13#discussion_r937034472. 

I thought splitting them felt right, but I don't feel strongly about it, especially since I'm not sure what the mime types will be used for. Like, I imagine that if we want to be able to filter the table for all images or all text, it's more efficient to do that if the top-level `media_type` is its own column, right?

On the other hand, splitting the mime type into two columns does seem like the _less_-obvious schema, so if we can't think of a compelling reason to do that, then keeping them as a single column sounds good to me.

Reply:
I think being able to filter on the `media_type` seems like a useful feature but are there use-cases of it that we would need for content creation or presentation?  Or is it more useful for analytics? If it's just analytics then I might suggest that we leave it as one top-level `mime-type` here because any analytics copy of this data can probably add the extra columns it needs to make analysis easier.

Reply:
I'm inclined to keep them separate, since it will make certain queries easier–like facet count by media_type for all images, as @kdmccormick points out, and we can always join them together for presentation purposes if we want.

Though this reminds me that I need to put proper indexing on them. :-/ I'll do that now...

Reply:
My assumption was that if you aren't indexing the fields anyways then it doesn't much matter if they're split or not - the performance will be similar because the DB still has to scan over all the values. On the other hand, if you use a single combined field but you _are_ indexing it, then prefix filtering like counting all the values that start with `image/` is essentially just as fast as if it's a unique value in its own column, because prefix matching with a text column index is extremely fast. (The only case where I see the split columns being useful would actually be if you needed to filter on the _second_ part of the media type without regard to the first part / prefix... but that doesn't seem to be a use case).

I also think that the prefixes aren't that meaningful in a lot of cases... Some might consider JSON to be a text value but it's in the `application/` prefix, and OLX is in neither `text/` nor `application/` as you might expect but `vnd/`. CSS and HTML, unlike JSON, are in the `text/` prefix; except XHTML (now deprecated) which is in `application/`. In short, the only prefix that I think is remotely useful or consistent is `image/`, and you can very easily and efficiently filter and count based on that prefix if there's an index on a combined text `media_type` column.

Anyhow I think either way is fine with me; I don't think it's a big deal at all, but I am also not convinced there's any benefit to splitting them up, while it is slightly more complex.

Reply:
> prefix matching with a text column index is extremely fast

Good to know. With that in mind, I'm fully in favor of a single field.

Reply:
Sorry, let me be more clear: any reservations I had about a single field are resolved. Two fields is also still OK with me. Agreed that this probably a very minor decision.

Reply:
Okay, I made it one field.